### ISSUE TITLE
Visual Studio solution for simulator TCTI

### ISSUE THREAD

**Issue by flihp** [tpm2-software/tpm2-tss#671]
This issue covers the creation of a solution file compatible with the free visual studio version capable of producing a dll and lib file for the simulator TCTI

**Comment by flihp** [comment]
Having spent some time on this in the last 2 weeks I've learned that the VS C compiler is not c99: https://visualstudio.uservoice.com/forums/121579-visual-studio-ide/suggestions/2089423-c99-support. General guidance from Microsoft is to use Visual Studio + LLVM / Clang. Apparently VS can use compilers other than the Microsoft VC++ one and if you need c99 support you must use either Clang or GCC, not VC++.

**Comment by AndreasFuchsSIT** [comment]
Am I correct in assuming that we are mostly lacking inttypes.h and stdint.h ?
Or are there other things we are missing of VS-C ?

There are quite some implementations of these for VS-C...

A BSD 3-clause one here for example: https://github.com/mattn/gntp-send/tree/master/include/msinttypes<https://github.com/mattn/gntp-send/blob/master/include/msinttypes/stdint.h>

We can put this into some directory of the project.
We can then have a feature-test of stdint.h in configure.ac, and conditionally add -I the file's directory to AM_CFLAGS.

Alternatively, for all we need with VS-C+x86/amd64 we could just write our own one. Depending on what we are missing.

A quick search gave me PRIx32 and PRIxPTR. Are there others we need ?

Cheers,
Andreas

________________________________

**Comment by flihp** [comment]
I've spent a pile of time doing battle with visual studio and the VC++ compiler trying to figure out answers to these sort of questions. I've run into 3 classes of problems in this work:
1) Issues like the one you point out here.
2) Issues with VC++ missing c99 support.
3) Microsoft / Windows specific APIs like winsock.

I've invested most of my time into the second category of issues and have come up with a solution: VC++ doesn't support c99 and Microsoft has stated publicly that they don't intend to support c99. So very few of our libraries can be built by VC++ but Microsoft has added support for various compiler backends to Visual Studio and the LLVM/Clang one works well. So the path forward on this is to use Clang, problem solved.

I'd been dealing with issues like the one you point out as I run into them. Your suggestion makes sense to me here and it's likely what we'll end up doing. The final issue will be handled in the build pulling in POSIX vs Windows code based on OS detection @ config time.
